#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#     libavPublisher.py
#
#     Copyright 2013 Dominique Hunziker
#
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#
#     \author/s: Dominique Hunziker
#
#

from dynamic_reconfigure.parameter_generator_catkin import *


def main():
    gen = ParameterGenerator()

    codecs = [gen.const('h264', str_t, 'h264', 'H264 (libx264)'),
              gen.const('ffv1', str_t, 'ffv1', 'Lossless frame by frame compression')]
    codec_enum = gen.enum(codecs, 'Enum to set the compression codec')

    pix_fmts = [gen.const('rgb', str_t, 'rgb', 'RGB 8:8:8'),
                gen.const('yuv', str_t, 'yuv', 'YUV 4:2:0'),
                gen.const('gray8', str_t, 'gray8', 'Single channel (8 bits)'),
                gen.const('gray16', str_t, 'gray16', 'Single channel (16 bits)')]
    pixel_enum = gen.enum(pix_fmts, 'Enum to set the pixel format of compressed stream')

    gen.add('codec', str_t, 0, 'Codec type (matching selected pixel format)', 'ffv1', edit_method=codec_enum)
    gen.add('pixel_format', str_t, 0, 'Pixel format (matching selected codec)', 'gray16', edit_method=pixel_enum)
    gen.add('encoded_width', int_t, 0, 'Image width in encoded state', 640)
    gen.add('encoded_height', int_t, 0, 'Image height in encoded state', 480)

    gen.add('gop_size', int_t, 0, 'GOP size (only for h264)', 150)
    gen.add('x264_preset', str_t, 0, 'Preset (only for h264)', 'fast')

    return gen.generate('libav_image_transport', 'LibAVImagePublisher',
                        'libavPublisher')


if __name__ == '__main__':
    exit(main())
